tests/test_common.py

```diff
@@ -326,6 +338,45 @@ def test_join(df_raw: Any) -> None:
     compare_dicts(result_native, expected)
 
 
+# Sample data
+df_pandas_left = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
+df_pandas_right = pd.DataFrame({"a": [1, 2, 3], "c": [4, 5, 6]})
+
+df_polars_left = pl.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
+df_polars_right = pl.DataFrame({"a": [1, 2, 3], "c": [4, 5, 6]})
+
+
+@pytest.mark.parametrize(
+    "df_left, df_right",
+    [(df_pandas_left, df_pandas_right), (df_polars_left, df_polars_right)],
+)
+def test_coalesce_overlapping_names(df_left, df_right):
+    if isinstance(df_left, pd.DataFrame):
+        # Convert Pandas to Polars for uniformity in processing
+        df_left = pl.DataFrame(df_left)
+        df_right = pl.DataFrame(df_right)
+
+    result = df_left.join(df_right, left_on="b", right_on="c", how="left")
+
+    if isinstance(df_left, pd.DataFrame):
+        result_native = result.to_pandas()
+    else:
+        result_native = result
+
+    if isinstance(result_native, pd.DataFrame):
+        result_native = result_native.rename(columns={"a_x": "a", "a_y": "a_right"})
+        result_dict = result_native.to_dict(orient="list")
+    else:
+        result_native = result_native.rename({"a": "a", "a_right": "a_right"}).to_pandas()
+        result_dict = result_native.to_dict(orient="list")
```

Comment:
as noted in https://github.com/narwhals-dev/narwhals/pull/312#discussion_r1648553370, we can't have logic in tests, this needs to go in `_pandas_like`

see https://github.com/narwhals-dev/narwhals/pull/312#discussion_r1648553370 for details

Reply:
Alright Marco.